### PR TITLE
feat: Add MLflow tracing, trace Assessment Sync, and Trace Export

### PR DESCRIPTION
**Pull request by Vperiodt**
## What this PR does / why we need it
Adds end-to-end MLflow integration for Feast: distributed tracing, trace export, assessment sync, and MLflow GenAI Dataset sync — enabling ML teams to trace agent-feature interactions, produce training data from production traces, and serve curated evaluation datasets through Feast's online/offline stores.
### 1. Distributed Tracing (`mlflow.enable_distributed_tracing: true`)
- **Server-side spans**: Every feature retrieval endpoint (`get-online-features`, `search`, `write-to-online-store`) creates an MLflow span with feature names, entity count, and project metadata.
- **Cross-process trace linking**: W3C `traceparent` header propagation — Feast spans become children of the caller's trace. MCP server instruments its internal httpx client and forwards `mcp-session-id` for session grouping.
- **In-process LLM span tagging**: A span processor (`feast_span_processor`) automatically stamps LLM/CHAT_MODEL spans with `feast.context_features` when the SDK is used directly via `feast_trace_scope()`.
- **PII redaction**: Optional `redact_entity_pii` scrubs entity values from span inputs.
- **Sampling**: `trace_sampling_ratio` controls what fraction of requests produce spans.
- **Graceful degradation**: Missing MLflow, incompatible versions, or span creation failures never block feature serving.
### 2. Trace Assessment Sync (`feast mlflow sync-assessments`)
- **Assessment ingestion**: Scans MLflow traces in an experiment for expectations and feedback, writes them to a Feast FeatureView or LabelView.
- **Flat mode** (default): Each assessment = one row with `trace_id`, `assessment_name`, `assessment_type`, `value`, `source_id`, `rationale`, `event_timestamp`.
- **Pivot mode** (`--pivot`): Groups assessments by `trace_id` into LabelView-compatible rows with `--assessment-mapping` to control column names and `--labeler-column` for the assessment source.
- **Selective sync**: `--assessment-names` filters which assessments to ingest.
- **LabelView support**: Falls back to `store.get_label_view()` when `store.get_feature_view()` fails, enabling correct offline writes via PushSource.
### 3. Trace Export (`feast mlflow export-traces`)
- **Trace extraction**: Queries MLflow for `CHAT_MODEL` spans, extracts prompt/completion pairs with `feast.*` context attributes and assessments (expectations + feedback).
- **Label resolution**: Two paths — join with a Feast LabelView (offline `ConflictPolicy` or online `LAST_WRITE_WINS`) or promote MLflow expectations to `corrected_response`.
- **Dataset filtering**: `--dataset` limits export to traces curated into a named MLflow GenAI Dataset.
- **Export formats**: `--format openai` (OpenAI-compatible JSONL) or `--format enriched` (includes trace provenance, feature refs, entity values, label info).
- **Artifact registration**: `--register` saves exported JSONL as an MLflow artifact; `--register-experiment` / `--register-run` control where.
```
Configuration:
   mlflow:
  enabled: true
  tracking_uri: "http://mlflow:5000"
  enable_distributed_tracing: true
  trace_sampling_ratio: 1.0
  redact_entity_pii: false
  ```
# Which issue(s) this PR fixes:

<p align="center">
  
  
</p>

<p align="center">
  
  
</p>

<p align="center">
  
  
</p>

# Checks
- [ ] I've made sure the tests are passing.
- [ ] My commits are signed off (`git commit -s`)
- [ ] My PR title follows [conventional commits](https://www.conventionalcommits.org/) format

## Testing Strategy
- [x] Unit tests
- [x] Integration tests
- [ ] Manual tests
- [ ] Testing is not required for this change

# Misc

---

<a href="https://app.devin.ai/review/feast-dev/feast/pull/6405" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

**Comment by franciscojavierarceo** [comment]
linter is failing

File: feast-tracing-demo/TRACING.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
+# Feast MLflow Tracing
```
Comment (ntkathole) [inline feast-tracing-demo/TRACING.md]: Add this official docs

**Review by jyejare — COMMENTED** [review]
This PR adds comprehensive MLflow distributed tracing support to Feast, enabling automatic trace collection for feature server endpoints, cross-process trace linking via traceparent headers, and in-process feature context tagging. The implementation is well-structured with proper separation of concerns, comprehensive test coverage, and graceful degradation when dependencies are missing.

File: sdk/python/feast/infra/online_stores/sqlite.py [hunk 2]
```diff
@@ -275,11 +275,14 @@ def online_read(
...
-                if ts.tzinfo is not None:
-                    res_ts = ts.astimezone(timezone.utc)
+                if isinstance(ts, (int, float)):
+                    res_ts = datetime.fromtimestamp(ts, tz=timezone.utc)
                 else:
-                    res_ts = ts.replace(tzinfo=timezone.utc)
+                    ts = cast(datetime, ts)
+                    if ts.tzinfo is not None:
+                        res_ts = ts.astimezone(timezone.utc)
+                    else:
```
Comment (jyejare) [inline sdk/python/feast/infra/online_stores/sqlite.py:421]: **[Warning] Inconsistent timestamp handling may cause runtime errors**

The code handles both int/float and datetime timestamp types, but the original code assumed datetime objects. This change could break existing code that expects consistent datetime handling and may cause timezone issues with numeric timestamps.

**Suggested:**
```suggestion
                ts = cast(datetime, ts)
                if ts.tzinfo is not None:
                    res_ts = ts.astimezone(timezone.utc)
                else:
                    res_ts = ts.replace(tzinfo=timezone.utc)
```

File: pyproject.toml [hunk 3]
```diff
@@ -105,6 +105,13 @@ mssql = ["ibis-framework[mssql]>=10.0.0"]
 oracle = ["ibis-framework[oracle]>=10.0.0"]
 mysql = ["pymysql", "types-PyMySQL"]
 openlineage = ["openlineage-python>=1.40.0"]
+mlflow = [
+    "mlflow>=2.14.0",
+    "opentelemetry-api>=1.28.0",
+    "opentelemetry-sdk>=1.28.0",
+    "opentelemetry-instrumentation-fastapi>=0.49b0",
+    "opentelemetry-instrumentation-httpx>=0.49b0",
```
Comment (jyejare) [inline pyproject.toml:119]: **[Suggestion] Consider adding opentelemetry-exporter-otlp-proto-http dependency**

The mlflow extra includes several OpenTelemetry packages but is missing opentelemetry-exporter-otlp-proto-http which was mentioned in the pixi.lock file. This could cause issues if OTLP export is needed.

**Suggested:**
```suggestion
mlflow = [
    "mlflow>=2.14.0",
    "opentelemetry-api>=1.28.0",
    "opentelemetry-sdk>=1.28.0",
    "opentelemetry-instrumentation-fastapi>=0.49b0",
    "opentelemetry-instrumentation-httpx>=0.49b0",
    "opentelemetry-exporter-otlp-proto-http>=1.28.0",
]
```

File: sdk/python/feast/feature_server.py [hunk 4]
```diff
@@ -300,6 +300,30 @@ async def load_static_artifacts(app: FastAPI, store):
...
+        from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
+
+        FastAPIInstrumentor.instrument_app(app)
+        logger.info("FastAPI OTEL instrumentation enabled for trace propagation")
+    except ImportError:
+        logger.debug(
+            "opentelemetry-instrumentation-fastapi not installed; "
+            "cross-process trace linking disabled"
+        )
+
```
Comment (jyejare) [inline sdk/python/feast/feature_server.py:653]: **[Suggestion] Improve error handling in FastAPI instrumentation**

The ImportError handling is good, but it would be helpful to also catch other potential exceptions during instrumentation and provide more specific error messages.

**Suggested:**
```suggestion
    try:
        from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor

        FastAPIInstrumentor.instrument_app(app)
        logger.info("FastAPI OTEL instrumentation enabled for trace propagation")
    except ImportError:
        logger.debug(
            "opentelemetry-instrumentation-fastapi not installed; "
            "cross-process trace linking disabled"
        )
    except Exception as e:
        logger.warning(f"Failed to instrument FastAPI app: {e}")
```

File: sdk/python/feast/tracing.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                for k, v in attributes.items():
+                    span.set_attribute(k, v)
+            yield span
+        # Flush before parent_ctx exits — the parent context cleanup
+        # removes the trace from MLflow's in-memory manager, so the
+        # async exporter must send the span before that happens.
+        if has_traceparent:
+            try:
+                _mlflow_mod.flush_trace_async_logging()
+            except Exception:
```
Comment (jyejare) [inline sdk/python/feast/tracing.py:202]: **[Warning] Potential performance impact from synchronous flush in async context**

Calling flush_trace_async_logging() synchronously in an async endpoint could block the event loop and impact performance. Consider making this operation truly asynchronous or moving it to a background task.

File: sdk/python/feast/tracing_context.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+@contextmanager
+def feast_trace_scope() -> Iterator[FeastTraceContext]:
+    """Context manager that creates and cleans up a ``FeastTraceContext``."""
+    ctx = FeastTraceContext()
+    _thread_local.feast_ctx = ctx
+    try:
+        yield ctx
+    finally:
+        ctx.clear()
```
Comment (jyejare) [inline sdk/python/feast/tracing_context.py:89]: **[Suggestion] Consider adding context validation**

The feast_trace_scope context manager could benefit from validation to ensure the context is properly cleaned up even if exceptions occur during setup.

**Suggested:**
```suggestion
@contextmanager
def feast_trace_scope() -> Iterator[FeastTraceContext]:
    """Context manager that creates and cleans up a ``FeastTraceContext``."""
    old_ctx = getattr(_thread_local, 'feast_ctx', None)
    ctx = FeastTraceContext()
    _thread_local.feast_ctx = ctx
    try:
        yield ctx
    finally:
        ctx.clear()
        _thread_local.feast_ctx = old_ctx
```

File: sdk/python/feast/mlflow_integration/config.py [hunk 7]
```diff
@@ -60,6 +60,12 @@ class MlflowConfig(FeastBaseModel):
     """ str: Suffix appended to the project name to form the MLflow
         experiment name for operation logs. Defaults to '-feast-ops'. """
 
+    enable_tracing: StrictBool = True
+    """ bool: When True and mlflow.enabled=True, server-side API calls
+        create MLflow trace spans via mlflow.start_span().  Spans appear
+        in the MLflow UI Traces tab and support parent-child linking via
+        traceparent headers.  Defaults to True. """
+
```
Comment (jyejare) [inline sdk/python/feast/mlflow_integration/config.py]: **[Nitpick] Consider more descriptive field name**

The field name 'enable_tracing' could be more specific since it only controls MLflow tracing, not all tracing in Feast.

**Suggested:**
```suggestion
    enable_distributed_tracing: StrictBool = True
    """ bool: When True and mlflow.enabled=True, server-side API calls
        create MLflow trace spans via mlflow.start_span().  Spans appear
        in the MLflow UI Traces tab and support parent-child linking via
        traceparent headers.  Defaults to True. """
```

File: pyproject.toml (hunk 3 above)
Comment (Vperiodt) [reply pyproject.toml:119]: Added that before it was causing grpcio conflicts and 
opentelemetry-exporter-otlp-proto-http is already a transitive dep of mlflow>=2.14.0

File: docs/reference/mlflow-distributed-tracing.md [hunk 8]
```diff
@@ -0,0 +1,132 @@
+# MLflow Distributed Tracing
```
Comment (ntkathole) [inline docs/reference/mlflow-distributed-tracing.md]: better to include this in existing mlflow.md

File: docs/SUMMARY.md [hunk 9]
```diff
@@ -196,6 +196,7 @@
 * [\[Alpha\] Feature View Versioning](reference/alpha-feature-view-versioning.md)
 * [OpenLineage Integration](reference/openlineage.md)
 * [MLflow Integration](reference/mlflow.md)
+* [MLflow Distributed Tracing](reference/mlflow-distributed-tracing.md)
```
Comment (ntkathole) [inline docs/SUMMARY.md]: no need of different section

File: sdk/python/feast/tracing_hooks.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+    attrs = ctx.get_context_attributes()
+    for key, value in attrs.items():
+        try:
+            span.set_attribute(key, value)
+        except Exception:
+            _logger.debug("Failed to set attribute %s on LLM span", key)
+
+
+def install_feast_span_processor() -> None:
```
Comment (ntkathole) [inline sdk/python/feast/tracing_hooks.py]: is it being called somewhere other than tests ?

Comment (Vperiodt) [reply sdk/python/feast/tracing_hooks.py]: It is expected to be called by the users in there agent code which helps to tag LLM spans with feast context, either i can document it or we can add it in the `_lazy_init()`

Comment (Vperiodt) [reply sdk/python/feast/tracing_hooks.py]: I updated it in the docs

File: sdk/python/feast/cli/finetuning.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
+"""CLI commands for fine-tuning dataset export."""
```
Comment (ntkathole) [inline sdk/python/feast/cli/finetuning.py]: feast finetuning is not correct naming here.

The feast finetuning name implies Feast has an opinion about LLM training workflows, which is a much larger product decision than what's in this PR.

I'ts better if command is like `feast mlflow export-traces` or similar - making it clear this is an MLflow integration subcommand

File: sdk/python/feast/cli/datasets.py [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+    click.echo(
+        f"Syncing MLflow dataset '{source}' → FeatureView '{feature_view}' "
+        f"({'full refresh' if full_refresh else 'incremental'})"
+    )
+    if tracking_uri:
+        click.echo(f"  MLflow tracking URI: {tracking_uri}")
+    if dry_run:
+        click.echo("  DRY RUN — no data will be written.")
+
+    result = sync_mlflow_dataset_to_feast(
```
Comment (ntkathole) [inline sdk/python/feast/cli/datasets.py]: I think better integration will be if it's `MlflowDatasetSource`. The dataset sync should be an MlflowDatasetSource (proper Feast primitive).

**Comment by ntkathole** [comment]
@Vperiodt Please rebase and solve conflicts, also please fix the linting

**Comment by ntkathole** [comment]
@Vperiodt conflicts
